Add Theatron to Architecture_Engine and Humans_Engine

rolyhudson opened the pull request:
Theatron is a set of Architecture objects for defining the geometry of stadia, theatre, cinemas, lecture halls and auditoria. It includes Human objects related to users of these spaces: Spectators, Audience, Eye, ViewCones and analytic methods for evaluating view quality and ergonomics

### NOTE: Depends on 
A fork of BHoM with object definitions in Architecture_oM and Humans_oM
pull request in BHoM : https://github.com/BHoM/BHoM/pull/529.
### Test files
Three zipped grasshopper files that demonstrate three key use case scenarios
[theatron-GH_Tests.zip](https://github.com/BHoM/BHoM_Engine/files/3418742/theatron-GH_Tests.zip)

### Additional comments
This is first time pull request to BHoM_Engine so its wrong, too complicated or missing something please let me know

FraserGreenroyd commented:
Hi @rolyhudson - welcome to the engine :smile:

First glance looks great, though I haven't done an in depth review (I'll save that for the morning!), so this is just my first thoughts :smile: I notice there's a conflict on the Environment Engine - namely, you've removed the `counter` variable which was removed some commits ago - I'm guessing you branched off a while back? We might need to resolve those to enable a smooth transition into `master` (once reviewed).

I imagine the review process will take a couple of days so we have plenty of time to work out a solution with @al-fisher as well :smile:

Once you've set up the PR in BHoM (which I think you have?), it's a good idea to link to it in this PR (under the `Depends On` header) so people can find it easily when it does come time to review :smile:

Other than that, I'm looking forward to doing a full review over the next couple of days :smile:

In hindsight... that's a lot of smiley faces... but I'm happy to look at new PRs with fresh stuff so I'm going to add another one here to cement and spread the happiness :smile:

rolyhudson commented:
Cheers @FraserGreenroyd  :smile:
I will take a look at the conflict on the Environment Engine

rolyhudson commented:
This has not fixed it!

FraserGreenroyd commented:
That looks to have done the trick @rolyhudson 😄

al-fisher commented:
This is awesome @rolyhudson - loving your work! 😄 
Overall this looks fantastic and the general approach looks really good. Excited to be building out the `Architecture` and `Human` namespaces.

As well as on first quick testing - got your sample scripts working pretty well too! 

![Stadium](https://user-images.githubusercontent.com/15924573/62134254-12d2dd80-b2d8-11e9-8cf8-2a47ee23dcea.gif)


![AValues ViewCones](https://user-images.githubusercontent.com/15924573/62131281-c89b2d80-b2d2-11e9-8b88-a1ec6ebe0afd.PNG)


I have noticed a few places in the code to consider tweaking before we merge - will raise a review with a few comments. 🚀

rolyhudson commented on Architecture_Engine/Create/Theatron/ActivityArea.cs:
```diff
@@ -0,0 +1,92 @@
+        /**** Public Methods                            ****/
+        /***************************************************/
+        [Description("Create an activityArea rectangle with width: 60 and length: 90, focal point is at 0,0,0")]
+        [Input("scale", "Optional input to scale default units from metres")]
```
I have many default values defined in metres in other objects so wondering what the current standard approach would be. Is there a specific example you could point me to?

Two more general but related thoughts:
on default values I imagine these should be in BHoM_Datasets?

I also have a set of classic stadium parameters that I imagine could be useful as a reference Dataset. AVIVA Dublin, Wembley, Aston Villa, Man City, Cardiff Millenium, Newcastle St James park, Hillsborough also these should be in BHoM_Datasets?

al-fisher replied:
The default values in meters you are defining across the code is exactly right. Same principle is being used in every other namespace. 
And definitely useful to have meaningful defaults - and must be in meters / (SI). So no problem there. My comment above was just stating that your scaling function can not change the units from meters in the core BHoM - hope that makes sense!

On Datasets for classic stadia - yep - exactly that! Great idea! Is super easy. You can use the File_Adapter or simply copy the `ToJson` serialised BHoM objects and place in a file in the Datasets folder. They will be reflected in automatically.

rolyhudson commented on Architecture_Engine/Compute/Theatron/CircularStadiaPlan.cs:
```diff
@@ -0,0 +1,62 @@
+
+namespace BH.Engine.Architecture.Theatron
+{
+    public static partial class Compute
```
OK I was a bit uncertain on those I'll switch them to Create

rolyhudson commented on Architecture_Engine/Create/Theatron/TheatronPlan.cs:
```diff
@@ -0,0 +1,179 @@
+        [Input("focalPolyline", "The polyline used to define focal points for Cvalue profile generation")]
+        public static TheatronPlan PlanGeometry(List<ProfileOrigin> structuralSections, Polyline focalPolyline)
+        {
+            var planGeometry = new TheatronPlan
```
:thumbsup:

rolyhudson commented on Architecture_Engine/Create/Theatron/TheatronFullProfile.cs:
```diff
@@ -0,0 +1,120 @@
+            Point origin = Geometry.Create.Point(minDist, 0, parameters[0].StartZ - parameters[0].EyePositionZ);
+            Vector direction = Vector.XAxis;
+            ProfileOrigin sectionOrigin = ProfileOrigin(origin, direction);
+            GenerateMapProfiles(ref fullProfile, parameters, minDist, sectionOrigin);
```
Thanks for that suggestion. Now fixed 👍

rolyhudson commented on Architecture_Engine/Create/Theatron/StadiaType.cs:
```diff
@@ -0,0 +1,52 @@
+        /***************************************************/
+        [Description("Create an enumerator for predfined stadia types.")]
+        [Input("typeNum", "0 = EightArc, 1 = Orthogonal, 2 = NoCorners, 3 = Circular, 4 = FourArc, 5 = Oblong, 6 = undefined")]
+        public static StadiaType StadiaTypeEnum(int typeNum)
```
Cool much easier 😃

rolyhudson commented:
> I added a couple of comments in the code that should help fix some issues.
@epignatelli question due to my ignorance:
Can I see all those comments in this conversation?

epignatelli commented:
> > I added a couple of comments in the code that should help fix some issues.
> 
> @epignatelli question due to my ignorance:
> Can I see all those comments in this conversation?

@rolyhudson sorry - I badly explained myself. With "in the code" I meant, "In the diff", not in the actual c# code. They're the one you just answered to.

rolyhudson commented:
> One more thing we can do in the future is to add a `Query.Geometry(TheatronGeometry theatron)` to trigger the automatic rendering of the object. Something like this should do:
> 
> ```cs
>         public static CompositeGeometry Geometry(this oM.Architecture.Theatron.TheatronGeometry theatron)
>         {
>             return Engine.Geometry.Create.CompositeGeometry(theatron?.Tiers3d?.SelectMany(x => x?.TierBlocks).Select(x => x?.Floor));
>         }
> ```

This is great. So anywhere I want to pull out resulting geometry without going through the explodes in GH/dynamo a standard BHoM approach would be add a method like this to the ```Architecture_Engine``` ```Query``` class?

epignatelli commented:
> 
> This is great. So anywhere I want to pull out resulting geometry without going through the explodes in GH/dynamo a standard BHoM approach would be add a method like this to the `Architecture_Engine` `Query` class?

Yes, precisely! The aim is two-fold: you can directly use the method in grasshopper to extract geometries, and it is called by the UIs to automatically render `BHoMObject`s.

rolyhudson commented:
> Yes, precisely! The aim is two-fold: you can directly use the method in grasshopper to extract geometries, and it is called by the UIs to automatically render `BHoMObject`s.

OK even better that its automatically rendering. Looking at where else this would be useful to implement. 

Can you explain why if the Floor property is a Mesh we use ```CompositeGeometry```?

And for example if I want to render each of the ```Profile``` property (a Polyline) of a list of ```TierProfile``` in ```TheatronFullProfile``` I assume I should also use ```CompositeGeometry```?
Like this:
```c#
public static CompositeGeometry Geometry(this oM.Architecture.Theatron.TheatronFullProfile theatronFullProfile)
        {
            return Engine.Geometry.Create.CompositeGeometry(theatronFullProfile?.BaseTierProfiles?.Select(x => x?.Profile));
        }
```

rolyhudson commented:
Addressed the changes and the related oM changes. 
I'm not sure of the formal next step but let me know what I've missed or should add. 
Thanks for reviewing this @al-fisher + @epignatelli 👍 👍 
Also I updated the GH demos:
[theatron GH demos.zip](https://github.com/BHoM/BHoM_Engine/files/3458821/theatron.GH.demos.zip)

al-fisher commented:
Fantastic - thanks @rolyhudson - will do final quick review since the commits just made. 
Process-wise - should all then be good, we'll look to merge to `master` 🚀 
If you have thoughts on further issues - to expand features, fix bugs etc. etc. do feel free to raise issues here in the BHoM org. 
And then if you want to tackle any, assign yourself and add to the SCRUM Project board: 
![image](https://user-images.githubusercontent.com/15924573/62354580-244cfd00-b505-11e9-8842-c9584a17e87c.png)

SCRUM Project Board: https://github.com/orgs/BHoM/projects/3  

Can branch or fork as you prefer. 
Cheers

epignatelli commented:
> 
> OK even better that its automatically rendering. Looking at where else this would be useful to implement.
> 
> Can you explain why if the Floor property is a Mesh we use `CompositeGeometry`?
> 
> And for example if I want to render each of the `Profile` property (a Polyline) of a list of `TierProfile` in `TheatronFullProfile` I assume I should also use `CompositeGeometry`?
> Like this:
> 
> ```cs
> public static CompositeGeometry Geometry(this oM.Architecture.Theatron.TheatronFullProfile theatronFullProfile)
>         {
>             return Engine.Geometry.Create.CompositeGeometry(theatronFullProfile?.BaseTierProfiles?.Select(x => x?.Profile));
>         }
> ```

So, the problem with `Floor` is that it actually is a `List<Mesh>`, am I correct?
The problem comes from the Grasshopper_Toolkit: https://github.com/BHoM/Grasshopper_Toolkit/blob/c404f2eca317818d0fc021ab79545659ef80a0af/Grasshopper_Engine/Objects/Types/GH_BHoMObject.cs#L212-L220

For each parameter, it stores a `IGeometry` object, and a `Geometry` method that returns a  `List<IGeometry>` will make the rendering fail right now, although it would still work as a standalone method.
But that's definitely a limitation that can be overcome!

Anyways, I just wanted to thumb up this pr for the high number of thumbs up, hearts, and rockets it yielded 😄

epignatelli commented:
@rolyhudson I just added some documentation about automatic rendering of objects in the wiki:
https://github.com/BHoM/documentation/wiki/BH.UI-%E2%80%90-Expose-Your-Code-to-UIs#automatic-rendering-of-a-bhomobject

Let me know if they're clear enough, and if not, feel free to hack that document!
